tests/nnstreamer_source/unittest_src_iio.cc

```diff
@@ -1564,6 +1564,86 @@ TEST (test_tensor_src_iio, data_verify_freq_generic_type)
   clean_iio_dev_structure (dev0);
 }
 
+/**
+ * @brief test the logic with invalid frequency value
+ */
+TEST (test_tensor_src_iio, set_frequency_n)
+{
+  iio_dev_dir_struct *dev0;
+  GstElement *src_iio_pipeline;
+  GstElement *src_iio;
+  GstStateChangeReturn status;
+  gchar *parse_launch;
+  gint data_value;
+  guint data_bits;
+  gint ret_freq;
+
+  data_value = DATA;
+  data_bits = 16;
+  /** Make device */
+  dev0 = make_full_device (data_value, data_bits);
+  ASSERT_NE (dev0, nullptr);
+  /** setup */
+  dev0->log_file = g_build_filename (dev0->base_dir, "temp.log", NULL);
+  parse_launch =
+      g_strdup_printf
+      ("%s iio-base-dir=%s device-number=%d name=my-src-iio ! fakesink",
+      ELEMENT_NAME, dev0->iio_base_dir_sim, 0);
+  src_iio_pipeline = gst_parse_launch (parse_launch, NULL);
+  ASSERT_NE (parse_launch, nullptr);
+  g_free (parse_launch);
+
+  /** get and verify the caps */
+  src_iio = gst_bin_get_by_name (GST_BIN (src_iio_pipeline), "my-src-iio");
+  ASSERT_NE (src_iio, nullptr);
+
+  /** set the invalid frequency */
+  g_object_set (src_iio, "frequency", -1, NULL);
+  g_object_get (src_iio, "frequency", &ret_freq, NULL);
+  EXPECT_EQ (ret_freq, -1);
+
+  /** fail with invalid frequency */
+  status = gst_element_set_state (src_iio_pipeline, GST_STATE_PLAYING);
+  EXPECT_NE (status, GST_STATE_CHANGE_ASYNC);
+
+  /** delete device structure */
+  gst_object_unref (src_iio);
+  gst_object_unref (src_iio_pipeline);
+}
+
+/**
+ * @brief test the logic with invalid base dir
+ */
+TEST (test_tensor_src_iio, set_base_dir_n)
+{
+  iio_dev_dir_struct *dev0;
+  GstHarness *hrnss = NULL;
+  GstElement *src_iio = NULL;
+  GstStateChangeReturn status;
+
+  /** Make device */
+  dev0 = make_full_device (DATA, 16);
+  ASSERT_NE (dev0, nullptr);
+
+  /** setup */
+  hrnss = gst_harness_new_empty ();
+  ASSERT_TRUE (hrnss != NULL);
+  gst_harness_add_parse (hrnss, ELEMENT_NAME);
+  src_iio = gst_harness_find_element (hrnss, ELEMENT_NAME);
+  ASSERT_TRUE (src_iio != NULL);
+
+  /** setup properties */
+  g_object_set (src_iio, "iio-base-dir", "/not/existed/path", NULL);
+  g_object_set (src_iio, "dev-dir", dev0->dev_dir, NULL);
+  g_object_set (src_iio, "device", DEVICE_NAME, NULL);
+
+  status = gst_element_set_state (src_iio, GST_STATE_PAUSED);
+  EXPECT_NE (status, GST_STATE_CHANGE_NO_PREROLL);
+
+  /** teardown */
+  gst_harness_teardown (hrnss);
```

Comment:
Need `gst_object_unref (src_iio);`?

Reply:
gst_harness_teardown appears to do it.
We may check if src_iio is available or not after LINE1644 in the code to make sure that it's freed. (don't just add unref there unless you sure that it's not freed at that point)

Reply:
I executed Valgrind to check mem leak. Unfortunately, there are some points which should be fixed. I'll make another PR that fixes the problem include this one.